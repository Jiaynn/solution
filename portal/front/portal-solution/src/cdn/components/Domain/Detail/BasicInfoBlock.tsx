
import React from 'react'
import classNames from 'classnames'
import { observer } from 'mobx-react'
import { chunk } from 'lodash'
import { useInjection } from 'qn-fe-core/di'
import Row from 'react-icecream/lib/row'
import Col from 'react-icecream/lib/col'
import { ICertInfo } from 'portal-base/certificate'
import { useTranslation } from 'portal-base/common/i18n'
import { UserInfoStore as UserInfo } from 'portal-base/user/account'

import { humanizeTimeUTC } from 'cdn/transforms/datetime'

import {
  humanizeOperatingState, humanizePlatform, humanizeGeoCover, humanizeType,
  humanizeRecycleLeftDays, humanizeIpTypes
} from 'cdn/transforms/domain/index'

import AbilityConfig from 'cdn/constants/ability-config'
import { isOEM, oemConfig } from 'cdn/constants/env'
import { DomainType, GeoCover } from 'cdn/constants/domain'

import HelpLink from 'cdn/components/common/HelpLink'

import TipIcon from 'cdn/components/TipIcon'

import { IDomainDetail } from 'cdn/apis/domain'

import BlockTitle from './BlockTitle'
import GeoCoverConfigBlock from './GeoCoverConfigBlock'
import IpTypesConfigBlock from './IpTypesConfigBlock'

export default observer(function BasicInfoBlock(props: {
  domain: IDomainDetail
  certInfo?: ICertInfo
  hasIcp: boolean
  onRefresh?: () => void
}) {
  const domain = props.domain
  const userInfoStore = useInjection(UserInfo)
  const abilityConfig = useInjection(AbilityConfig)
  const t = useTranslation()
  const isTestDomain = domain.type === DomainType.Test
  const showCname = !isTestDomain && !(
    isOEM && oemConfig.hideSubAccountDomainDetailCname && userInfoStore.parent_uid !== 0
  )

  const cnameBlock = showCname && (
    <>
      <span className="key">CNAME：</span>
      <span className="value">
        {domain.cname}
        <HelpLink
          className="cname-help-link"
          href="https://developer.qiniu.com/fusion/kb/1322/how-to-configure-cname-domain-name"
        >
          帮助
        </HelpLink>
      </span>
    </>
  )

  const createAtBlock = (
    <>
      <span className="key">创建时间：</span>
      <span className="value">{humanizeTimeUTC(domain.createAt)}</span>
    </>
  )

  const domainStateBlock = (
    <>
      <span className="key">域名状态：</span>
      <span className={classNames(['value', `state-${domain.operatingState}`])}>
        {t(humanizeOperatingState(domain.operatingState, domain.freezeType))}
      </span>
    </>
  )

  const leftDaysBlock = isTestDomain && (
    <>
      <span className="key">剩余回收时间：</span>
      <span className="value text-danger">
        {humanizeRecycleLeftDays(domain.leftDays!)}
      </span>
    </>
  )

  const domainTypeBlock = (
    <>
      <span className="key">域名类型：</span>
      <span className="value">{t(humanizeType(domain.type))}</span>
    </>
  )

  const protocolBlock = (
    <>
      <span className="key">协议：</span>
      <span className="value">{domain.protocol.toUpperCase()}</span>
    </>
  )

  const platformBlock = !abilityConfig.hideDomainPlatform && (
    <>
      <span className="key">使用场景：</span>
      <span className="value">{t(humanizePlatform(domain.platform))}</span>
    </>
  )

  const geoCoverBlock = (
    <>
      <span className="key">覆盖：</span>
      <span className="value">{t(humanizeGeoCover(domain.geoCover))}</span>
      <GeoCoverConfigBlock
        onConfigOk={props.onRefresh!}
        hasIcp={props.hasIcp}
        domain={domain}
        certInfo={props.certInfo}
      />
    </>
  )

  const ipTypesTip = domain.geoCover === GeoCover.Foreign
    && <TipIcon tip="海外暂时不支持 IPv6 访问" />
  const ipTypesBlock = (
    <>
      <span className="key">IP 协议{ipTypesTip}：</span>
      <span className="value">{humanizeIpTypes(domain.ipTypes)}</span>
      <IpTypesConfigBlock
        onConfigOk={props.onRefresh!}
        domain={domain}
        certInfo={props.certInfo}
      />
    </>
  )

  const blockItems = [
    cnameBlock, domainStateBlock, leftDaysBlock, protocolBlock, geoCoverBlock,
    createAtBlock, domainTypeBlock, platformBlock, ipTypesBlock
  ].filter(Boolean)

  return (
    <section className="content-block">
      <BlockTitle>基本信息</BlockTitle>
      {chunk(blockItems, 4).map((rowItems, rowIndex) => (
        <Row className="content-line" gutter={8} key={rowIndex}>
          {rowItems.map((item, index) => {
            let colSpan = 6
            // 含有 cname 时，第一列宽度统一调长
            if (showCname) {
              colSpan = index === 0 ? 8 : 5
            }
            return <Col key={index} span={colSpan}>{item}</Col>
          })}
        </Row>
      ))}
    </section>
  )
})
