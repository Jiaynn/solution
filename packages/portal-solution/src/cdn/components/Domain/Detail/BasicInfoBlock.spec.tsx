/*
 * @file cases for domain create component
 * @author nighca <nighca@live.cn>
 */

import React from 'react'

import { RendererUtils as Renderer } from 'test'

import mockDomainDetail from 'cdn/test/domain-detail-mock'

import BasicInfoBlock from './BasicInfoBlock'

const domain = mockDomainDetail()

const noop = () => {}

it('renders correctly', () => {
  const renderer = new Renderer()
  const tree = renderer.createWithAct(
    <BasicInfoBlock domain={domain} hasIcp onRefresh={noop} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
