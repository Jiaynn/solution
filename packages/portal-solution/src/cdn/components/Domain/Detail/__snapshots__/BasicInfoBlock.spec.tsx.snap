// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="content-block"
>
  <h4
    className="content-block-title"
  >
    基本信息
  </h4>
  <div
    className="ant-row content-line"
    style={
      Object {
        "marginLeft": -4,
        "marginRight": -4,
      }
    }
  >
    <div
      className="ant-col ant-col-8"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        CNAME：
      </span>
      <span
        className="value"
      >
        10rduchl.qiniudns.com
        <a
          className="cname-help-link"
          href="https://developer.qiniu.com/fusion/kb/1322/how-to-configure-cname-domain-name"
          rel="noopener"
          target="_blank"
        >
          帮助
        </a>
      </span>
    </div>
    <div
      className="ant-col ant-col-5"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        域名状态：
      </span>
      <span
        className="value state-success"
      >
        成功
      </span>
    </div>
    <div
      className="ant-col ant-col-5"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        协议：
      </span>
      <span
        className="value"
      >
        HTTPS
      </span>
    </div>
    <div
      className="ant-col ant-col-5"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        覆盖：
      </span>
      <span
        className="value"
      >
        中国大陆
      </span>
      <button
        className="ant-btn basic-block-btn ant-btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span>
          修改
        </span>
      </button>
    </div>
  </div>
  <div
    className="ant-row content-line"
    style={
      Object {
        "marginLeft": -4,
        "marginRight": -4,
      }
    }
  >
    <div
      className="ant-col ant-col-8"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        创建时间：
      </span>
      <span
        className="value"
      >
        2016-03-19 15:26:15
      </span>
    </div>
    <div
      className="ant-col ant-col-5"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        域名类型：
      </span>
      <span
        className="value"
      >
        普通域名
      </span>
    </div>
    <div
      className="ant-col ant-col-5"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        使用场景：
      </span>
      <span
        className="value"
      >
        图片小文件
      </span>
    </div>
    <div
      className="ant-col ant-col-5"
      style={
        Object {
          "paddingLeft": 4,
          "paddingRight": 4,
        }
      }
    >
      <span
        className="key"
      >
        IP 协议
        ：
      </span>
      <span
        className="value"
      >
        IPv4 / IPv6
      </span>
      <button
        className="ant-btn basic-block-btn ant-btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span>
          修改
        </span>
      </button>
    </div>
  </div>
</section>
`;
